c2rust-analyze/src/rewrite/mod.rs

```diff
@@ -0,0 +1,231 @@
+use rustc_hir::Mutability;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Span;
+use std::fmt;
+
+mod apply;
+mod expr;
+mod span_index;
+mod ty;
+
+pub use self::expr::gen_expr_rewrites;
+pub use self::ty::dump_rewritten_local_tys;
+pub use self::ty::gen_ty_rewrites;
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Rewrite<S = Span> {
+    /// Take the original expression unchanged.
+    Identity,
+    /// Extract the subexpression at the given index.
+    Sub(usize, S),
+
+    // Expression builders
+    /// `&e`, `&mut e`
+    Ref(Box<Rewrite>, Mutability),
+    /// `core::ptr::addr_of!(e)`, `core::ptr::addr_of_mut!(e)`
+    AddrOf(Box<Rewrite>, Mutability),
+    /// `*e`
+    Deref(Box<Rewrite>),
+    /// `arr[idx]`
+    Index(Box<Rewrite>, Box<Rewrite>),
+    /// `arr[idx..]`
+    SliceTail(Box<Rewrite>, Box<Rewrite>),
+    /// `e as usize`
+    CastUsize(Box<Rewrite>),
+    /// The integer literal `0`.
+    LitZero,
+
+    // Type builders
+    /// Emit a complete pretty-printed type, discarding the original annotation.
+    PrintTy(String),
+    /// `*const T`, `*mut T`
+    TyPtr(Box<Rewrite>, Mutability),
+    /// `&T`, `&mut T`
+    TyRef(Box<Rewrite>, Mutability),
+    /// `[T]`
+    TySlice(Box<Rewrite>),
+    /// `Foo<T1, T2>`
+    TyCtor(String, Vec<Rewrite>),
+}
+
+impl Rewrite {
+    fn pretty(&self, f: &mut fmt::Formatter, prec: usize) -> fmt::Result {
+        fn parenthesize_if(
+            cond: bool,
+            f: &mut fmt::Formatter,
+            inner: impl FnOnce(&mut fmt::Formatter) -> fmt::Result,
+        ) -> fmt::Result {
+            if cond {
+                f.write_str("(")?;
+            }
+            inner(f)?;
+            if cond {
+                f.write_str(")")?;
+            }
+            Ok(())
+        }
+
+        // Expr precedence:
+        // - Index, SliceTail: 3
+        // - Ref, Deref: 2
+        // - CastUsize: 1
+        //
+        // Currently, we don't have any type builders that require parenthesization.
+
+        match *self {
+            Rewrite::Identity => write!(f, "$e"),
+            Rewrite::Sub(i, _) => write!(f, "${}", i),
```

Comment:
What do the `$e` and `${}` mean?

Reply:
The syntax here is arbitrary since it's only used for debug prints.  I picked `$e` (like a macro placeholder) to represent the whole original expression, and `$0`, `$1`, etc for the subexpressions of the original expression.

Reply:
Oh okay, I didn't realize it was only for debug prints.

Reply:
Added a comment.